### PR TITLE
[Tests] Simplify MefTestHelper

### PR DESCRIPTION
**Pull request by duncanp-sonar**
- removed unused code
- added overload to simplify calling code

No product code changes - test only.

### CODE REVIEW

**Comment by duncanp-sonar** [comment]
There are a lot of files in this changed, but they are mostly boilerplate. I'll explicitly all of non-boilerplate changes.

### Summary
* there are no product code changes.
* I removed unnecessary code from `MefTestHelpers` - both unused methods, and unused arguments. For example, we don't use MEF optional imports and were always passing `null`.
* I added a `CreateExport` override to simplify the calling code.



Previously, the standard MEF test looked like this:

```
        [TestMethod]
        public void MefCtor_CheckIsExported()
        {
            MefTestHelpers.CheckTypeCanBeImported<CMakeRequestFactory, IRequestFactory>(null, new[]
            {
                MefTestHelpers.CreateExport<ICompilationConfigProvider>(Mock.Of<ICompilationConfigProvider>()),
                MefTestHelpers.CreateExport<ICFamilyRulesConfigProvider>(Mock.Of<ICFamilyRulesConfigProvider>()),
                MefTestHelpers.CreateExport<IEnvironmentVarsProvider>(Mock.Of<IEnvironmentVarsProvider>())
            });
        }
```	

Now, they look like this:

```		
		[TestMethod]
        public void MefCtor_CheckIsExported()
        {
            MefTestHelpers.CheckTypeCanBeImported<CMakeRequestFactory, IRequestFactory>(
                MefTestHelpers.CreateExport<ICompilationConfigProvider>(),
                MefTestHelpers.CreateExport<ICFamilyRulesConfigProvider>(),
                MefTestHelpers.CreateExport<IEnvironmentVarsProvider>());
        }
```	
		
		
While making this change I discovered another bug in the MEF tests (a case where tests pass when they should fail). I'll handle that in a separate PR.

File: src/Infrastructure.VS.UnitTests/LanguageDetection/SonarLanguageRecognizerTests.cs [hunk 1]
```diff
@@ -55,18 +55,11 @@ public void TestInitialize()
         }
 
         [TestMethod]
-        public void Class_IsMEF()
```
Comment (duncanp-sonar) [inline src/Infrastructure.VS.UnitTests/LanguageDetection/SonarLanguageRecognizerTests.cs:58]: Changed the test to the standard format.

File: src/Integration.UnitTests/Helpers/SonarLintOutputTests.cs [hunk 2]
```diff
@@ -43,12 +43,8 @@ public void Ctor_WithNullSettings_ThrowsArgumentNullException()
         [TestMethod]
         public void MefCtor_CheckIsExported()
         {
-            // Arrange
```
Comment (duncanp-sonar) [inline src/Integration.UnitTests/Helpers/SonarLintOutputTests.cs:46]: Changed the test to the standard format.

File: src/Integration.UnitTests/LocalServices/CredentialStoreTests.cs [hunk 3]
```diff
@@ -51,11 +51,8 @@ public void TestInitialize()
         [TestMethod]
         public void MefCtor_CheckIsExported()
         {
-            // Arrange
```
Comment (duncanp-sonar) [inline src/Integration.UnitTests/LocalServices/CredentialStoreTests.cs:54]: Changed the test to the standard format.

File: src/Integration.UnitTests/NewConnectedMode/ConfigurationProviderTests.cs [hunk 4]
```diff
@@ -55,16 +55,10 @@ public void TestInitialize()
         [TestMethod]
         public void MefCtor_CheckIsExported()
         {
-            // Arrange
```
Comment (duncanp-sonar) [inline src/Integration.UnitTests/NewConnectedMode/ConfigurationProviderTests.cs:58]: Changed the test to the standard format.

File: src/IssueViz.Security.UnitTests/OpenInIDE/Http/HttpListenerFactoryTests.cs [hunk 5]
```diff
@@ -40,12 +40,9 @@ public class HttpListenerFactoryTests
         [TestMethod]
         public void MefCtor_CheckIsExported()
         {
-            // Arrange
```
Comment (duncanp-sonar) [inline src/IssueViz.Security.UnitTests/OpenInIDE/Http/HttpListenerFactoryTests.cs:43]: Changed the test to the standard format.

File: src/IssueViz.Security.UnitTests/OpenInIDE/Http/OwinPipelineProcessorTests.cs [hunk 6]
```diff
@@ -40,12 +40,9 @@ public class OwinPipelineProcessorTests
         [TestMethod]
         public void MefCtor_CheckIsExported()
         {
-            // Arrange
```
Comment (duncanp-sonar) [inline src/IssueViz.Security.UnitTests/OpenInIDE/Http/OwinPipelineProcessorTests.cs:43]: Changed the test to the standard format.

File: src/IssueViz.Security.UnitTests/OpenInIDE/Http/ShowHotspotOwinRequestHandlerTests.cs [hunk 7]
```diff
@@ -37,12 +37,9 @@ public class ShowHotspotOwinRequestHandlerTests
         [TestMethod]
         public void MefCtor_CheckIsExported()
         {
-            // Arrange
```
Comment (duncanp-sonar) [inline src/IssueViz.Security.UnitTests/OpenInIDE/Http/ShowHotspotOwinRequestHandlerTests.cs:40]: Changed the test to the standard format.

File: src/IssueViz.UnitTests/Editor/LocationNavigatorTests.cs [hunk 8]
```diff
@@ -61,18 +61,10 @@ public void TestInitialize()
         [TestMethod]
         public void MefCtor_CheckIsExported()
         {
-            // Arrange
```
Comment (duncanp-sonar) [inline src/IssueViz.UnitTests/Editor/LocationNavigatorTests.cs:64]: Changed the test to the standard format.

File: src/IssueViz.UnitTests/Editor/LocationTagging/LocationTaggerProviderTests.cs [hunk 9]
```diff
@@ -41,15 +41,11 @@ public class LocationTaggerProviderTests : CommonTaggerProviderTestsBase
         [TestMethod]
         public void MefCtor_CheckIsExported()
         {
-            var storeExport = MefTestHelpers.CreateExport<IIssueLocationStoreAggregator>(ValidLocationStore);
```
Comment (duncanp-sonar) [inline src/IssueViz.UnitTests/Editor/LocationTagging/LocationTaggerProviderTests.cs:44]: Changed the test to the standard format.

File: src/IssueViz.UnitTests/Editor/SelectedIssueTagging/Adornment/IssueLocationAdornmentTaggerProviderTests.cs [hunk 10]
```diff
@@ -36,14 +36,9 @@ public class IssueLocationAdornmentTaggerProviderTests : CommonViewTaggerProvide
         [TestMethod]
         public void MefCtor_CheckIsExported()
         {
-            var aggregatorFactoryExport = MefTestHelpers.CreateExport<IBufferTagAggregatorFactoryService>(Mock.Of<IBufferTagAggregatorFactoryService>());
```
Comment (duncanp-sonar) [inline src/IssueViz.UnitTests/Editor/SelectedIssueTagging/Adornment/IssueLocationAdornmentTaggerProviderTests.cs:39]: Changed the test to the standard format.

File: src/IssueViz.UnitTests/Editor/SelectedIssueTagging/Buffer/SelectedIssueLocationTaggerProviderTests.cs [hunk 11]
```diff
@@ -36,14 +36,10 @@ public class SelectedIssueLocationTaggerProviderTests : CommonTaggerProviderTest
         [TestMethod]
         public void MefCtor_CheckIsExported()
         {
-            var aggregatorFactoryExport = MefTestHelpers.CreateExport<IBufferTagAggregatorFactoryService>(Mock.Of<IBufferTagAggregatorFactoryService>());
```
Comment (duncanp-sonar) [inline src/IssueViz.UnitTests/Editor/SelectedIssueTagging/Buffer/SelectedIssueLocationTaggerProviderTests.cs:39]: Changed the test to the standard format.

File: src/IssueViz.UnitTests/Editor/SelectedIssueTagging/Highlighting/IssueHighlightViewTaggerProviderTests.cs [hunk 12]
```diff
@@ -36,15 +36,9 @@ public class IssueHighlightViewTaggerProviderTests : CommonViewTaggerProviderTes
         [TestMethod]
         public void MefCtor_CheckIsExported()
         {
-            var aggregatorFactoryExport = MefTestHelpers.CreateExport<IBufferTagAggregatorFactoryService>(Mock.Of<IBufferTagAggregatorFactoryService>());
```
Comment (duncanp-sonar) [inline src/IssueViz.UnitTests/Editor/SelectedIssueTagging/Highlighting/IssueHighlightViewTaggerProviderTests.cs:39]: Changed the test to the standard format.

File: src/IssueViz.UnitTests/Selection/IssueFlowStepNavigatorTests.cs [hunk 13]
```diff
@@ -49,14 +49,9 @@ public void TesInitialize()
         [TestMethod]
         public void MefCtor_CheckIsExported()
         {
-            var selectionServiceExport = MefTestHelpers.CreateExport<IAnalysisIssueSelectionService>(Mock.Of<IAnalysisIssueSelectionService>());
```
Comment (duncanp-sonar) [inline src/IssueViz.UnitTests/Selection/IssueFlowStepNavigatorTests.cs:52]: Changed the test to the standard format.

File: src/IssueViz.UnitTests/TableControls/VSTableEventProcessorProviderTests.cs [hunk 14]
```diff
@@ -34,12 +34,8 @@ public class VSTableEventProcessorProviderTests
         [TestMethod]
         public void MefCtor_CheckIsExported()
         {
-            // Arrange
```
Comment (duncanp-sonar) [inline src/IssueViz.UnitTests/TableControls/VSTableEventProcessorProviderTests.cs:37]: Changed the test to the standard format.

File: src/TestInfrastructure/Helpers/MefTestHelpers.cs [hunk 15]
```diff
@@ -24,28 +24,24 @@
...
 using System.ComponentModel.Composition.Primitives;
 using System.Linq;
+using Moq;
 
 namespace SonarLint.VisualStudio.Integration.UnitTests
 {
     public static class MefTestHelpers
     {
-        public static Export CreateExport<T>(object exportInstance, string contractName = null)
+        public static Export CreateExport<T>() where T: class
```
Comment (duncanp-sonar) [inline src/TestInfrastructure/Helpers/MefTestHelpers.cs:33]: New overload for when we just want a simple Mock of the exported type (which is most of the time).

File: src/TestInfrastructure/Helpers/MefTestHelpers.cs [hunk 16]
```diff
@@ -24,28 +24,24 @@
...
-        public static Export CreateExportWithMetadata<T>(object exportInstance, string metadataKey, string metadataValue)
-        {
-            Dictionary<string, object> metadata = new Dictionary<string, object>();
-            metadata[metadataKey] = metadataValue;
-
-            return CreateExport<T>(exportInstance, metadata);
+            return CreateExport<T>(Mock.Of<T>());
         }
 
-        public static Export CreateExport<T>(object exportInstance, IDictionary<string, object> metadata, string contractName = null)
```
Comment (duncanp-sonar) [inline src/TestInfrastructure/Helpers/MefTestHelpers.cs:45]: We don't use MEF metadata, so I removed those arguments from the overloads.

File: src/TestInfrastructure/Helpers/MefTestHelpers.cs [hunk 17]
```diff
@@ -61,14 +57,13 @@ public static Export CreateExport<T>(object exportInstance, IDictionary<string,
         /// If it does use an explicit contract name you will need to call the other overload and supply an
         /// importer with a property with the appropriate [Import("MyContractName")] attribute.</remarks>
         public static SingleObjectImporter<TImportType> CheckTypeCanBeImported<TTypeToCheck, TImportType>(
-            IEnumerable<Export> optionalExports,
```
Comment (duncanp-sonar) [inline src/TestInfrastructure/Helpers/MefTestHelpers.cs:64]: We didn't use optional exports either.

File: src/TestInfrastructure/Helpers/MefTestHelpers.cs [hunk 18]
```diff
@@ -84,23 +79,23 @@ public static Export CreateExport<T>(object exportInstance, IDictionary<string,
         /// <typeparam name="TTypeToCheck">The type to check</typeparam>
         /// <typeparam name="TImportType">The import contract type</typeparam>
         /// <param name="importer">The instance that is importing values</param>
-        /// <param name="optionalExports">Any optional exports to include in the composition.</param>
         /// <param name="requiredExports">Any exports required by TTypeToCheck. Optional.</param>
-        public static void CheckTypeCanBeImported<TTypeToCheck, TImportType>(
+        private static void CheckTypeCanBeImported<TTypeToCheck, TImportType>(
```
Comment (duncanp-sonar) [inline src/TestInfrastructure/Helpers/MefTestHelpers.cs:83]: Not called externally -> can be `private`

File: src/TestInfrastructure/Helpers/MefTestHelpers.cs [hunk 19]
```diff
@@ -84,23 +79,23 @@ public static Export CreateExport<T>(object exportInstance, IDictionary<string,
...
             object importer,
-            IEnumerable<Export> optionalExports,
             IEnumerable<Export> requiredExports)
             where TImportType : class
             where TTypeToCheck : class
         {
-            if (requiredExports == null)
+            if (importer == null)
```
Comment (duncanp-sonar) [inline src/TestInfrastructure/Helpers/MefTestHelpers.cs:89]: Additional sanity check - should never be null.

File: src/Integration.Vsix.UnitTests/CFamily/VcxRequestFactoryTests.cs [hunk 20]
```diff
@@ -47,12 +47,10 @@ public class VcxRequestFactoryTests
...
         public void MefCtor_CheckIsExported()
         {
-            MefTestHelpers.CheckTypeCanBeImported<VcxRequestFactory, IRequestFactory>(null, new[]
-            {
-                MefTestHelpers.CreateExport<VsShell.SVsServiceProvider>(Mock.Of<IServiceProvider>()),
-                MefTestHelpers.CreateExport<ICFamilyRulesConfigProvider>(Mock.Of<ICFamilyRulesConfigProvider>()),
-                MefTestHelpers.CreateExport<ILogger>(Mock.Of<ILogger>())
-            });
+            MefTestHelpers.CheckTypeCanBeImported<VcxRequestFactory, IRequestFactory>(
+                MefTestHelpers.CreateExport<VsShell.SVsServiceProvider>(),
```
Comment (rita-sonarsource) [inline src/Integration.Vsix.UnitTests/CFamily/VcxRequestFactoryTests.cs:51]: In other places we explicitly passed Mock<IServiceProvider>. We don't need it?

File: src/IssueViz.UnitTests/Editor/DocumentNavigatorTests.cs [hunk 21]
```diff
@@ -73,13 +73,11 @@ public void TestInitialize()
...
         {
-            MefTestHelpers.CheckTypeCanBeImported<DocumentNavigator, IDocumentNavigator>(null, new[]
-            {
-                MefTestHelpers.CreateExport<SVsServiceProvider>(Mock.Of<IServiceProvider>()),
-                MefTestHelpers.CreateExport<IVsEditorAdaptersFactoryService>(Mock.Of<IVsEditorAdaptersFactoryService>()),
-                MefTestHelpers.CreateExport<IOutliningManagerService>(Mock.Of<IOutliningManagerService>()),
-                MefTestHelpers.CreateExport<IEditorOperationsFactoryService>(Mock.Of<IEditorOperationsFactoryService>())
-            });
+            MefTestHelpers.CheckTypeCanBeImported<DocumentNavigator, IDocumentNavigator>(
+                MefTestHelpers.CreateExport<SVsServiceProvider>(),
```
Comment (rita-sonarsource) [inline src/IssueViz.UnitTests/Editor/DocumentNavigatorTests.cs:77]: Same comment about IServiceProvider

**Review by rita-sonarsource — APPROVED** [review]
LGTM, just a comment that some tests do
`MefTestHelpers.CreateExport<SVsServiceProvider>()`
and others
`MefTestHelpers.CreateExport<SVsServiceProvider>(Mock.Of<IServiceProvider>())`

If the first works then it would be simpler to use just that.

**Pull request merged**